nibabel/funcs.py

```diff
@@ -115,15 +117,19 @@ def concat_images(images, check_affines=True):
     affine = img0.affine
     header = img0.header
     out_shape = (n_imgs, ) + i0shape
```

Comment:
This loses the memory efficiency of the original.  How about:

```
if axis is None:  # collect images in output array for efficiency
    out_shape = (n_imgs, ) + i0shape
    out_data = np.empty(out_shape)
else:  # collect images in list for use with np.concatenate
    out_data = [None] * n_imgs
```

Then you can use `out_data[i] = img.get_data()` in the loop.